app/scripts/controllers/transactions/tx-state-manager.js

```diff
@@ -187,28 +187,43 @@ export default class TransactionStateManager extends EventEmitter {
     const transactions = this.getTransactions({
       filterToCurrentNetwork: false,
     });
-    const txCount = transactions.length;
     const { txHistoryLimit } = this;
 
     // checks if the length of the tx history is longer then desired persistence
     // limit and then if it is removes the oldest confirmed or rejected tx.
     // Pending or unapproved transactions will not be removed by this
-    // operation.
+    // operation. For safety of presenting a fully functional transaction UI
+    // representation, this function will not break apart transactions with the
+    // same nonce, per network. Not accounting for transactions of the same
+    // nonce and network combo can result in confusing or broken experiences
+    // in the UI.
     //
     // TODO: we are already limiting what we send to the UI, and in the future
     // we will send UI only collected groups of transactions *per page* so at
     // some point in the future, this persistence limit can be adjusted. When
     // we do that I think we should figure out a better storage solution for
     // transaction history entries.
-    if (txCount > txHistoryLimit - 1) {
-      const index = transactions.findIndex((metaTx) => {
-        return getFinalStates().includes(metaTx.status);
-      });
-      if (index !== -1) {
-        this._deleteTransaction(transactions[index].id);
-      }
-    }
+    const nonceNetworkSet = new Set();
```

Comment:
I want to make sure I understand this.

So the goal is to ensure that if there are two or more transactions with the same nonce in the transaction list, that they don't get deleted.

So we iterate through the transactions. If a transaction is either counted as below `txHistoryLimit - 1` or is not in a final state, then it is NOT deleted. If a transaction is counted higher than `txHistoryLimit - 1` or is in a final state BUT has the same nonce as a transaction already flagged as one that shouldn't be deleted, then it is NOT deleted. If a transaction is above the txHistoryLimit count or is in a final state and does not match the nonce of a transaction already identified as one that shouldn't be deleted, then it is deleted...

So something about this isn't adding up for me... probably me misunderstanding, but I want to check.

Suppose that there are only 10 transactions and they are all "final states". Wouldn't they all be deleted according to this logic?

Should the `if (txCount > txHistoryLimit - 1) {` that was removed above wrap this tx deletion code?

And this is in a function that only adds a single tx. Before it only removed a single tx. Why now attempt to remove multiple within this function?

Reply:
maybe I forget how `filter` works...

Reply:
maybe not

Reply:
Based on my reading of the tests I think I did misunderstand the goal though

Reply:
Can the purpose of this code change be stated as:

If the addition of a transaction brings the transaction count above the txHistoryLimit then all transactions that share a nonce with the oldest "finalized" transaction should be deleted.

?

Reply:
>So the goal is to ensure that if there are two or more transactions with the same nonce in the transaction list, that they don't get deleted.

It's to ensure that one transaction in a group isn't deleted without deleting the entire group.

We can't let _all_ groups of more than 1 transaction persist indefinitely - then we'd have no limit on storage. But we can't break groups apart either without breaking the UI.

So the solution we went with here was to interpret the transaction history limit of 50 as meaning 50 _groups of transactions_. If there are over 50 groups, we delete any groups beyond that 50 (which should always be a single group if this is the only means of adding transactions, but might be more than 1 group if transactions are added elsewhere).

Reply:
when we return false from this method we are *Not* including it, when we return true we *are* including it. So when return true we are saying "this is an item we want to delete". if there are only ten transactions non will be deleted because they will never hit the limit and thus will all be added to the nonceSet

Reply:
basically yes as stated that will work